woodwork/datatable.py

```diff
@@ -524,16 +527,21 @@ def set_types(self, logical_types=None, semantic_tags=None, retain_index_tags=Tr
 
         new_dt = self._new_dt_from_cols(self._dataframe.columns)
         cols_to_update = {}
+        reset_index = False
         for col_name, col in new_dt.columns.items():
             if col_name not in logical_types and col_name not in semantic_tags:
                 continue
+            if 'index' in col.semantic_tags and not retain_index_tags:
+                reset_index = True
             if col_name in logical_types:
                 col = col.set_logical_type(logical_types[col_name], retain_index_tags)
             if col_name in semantic_tags:
                 col = col.set_semantic_tags(semantic_tags[col_name], retain_index_tags)
             cols_to_update[col_name] = col
 
         new_dt._update_columns(cols_to_update)
+        if reset_index:
```

Comment:
Another option for this check is to have `if new_dt.index is None and self.index is not None`. Plus this check could be used here, `DataTable.drop`, and the other semantic tag update methods, since all of them return a new DataTable, so it might be nice to have these lines be similar between all of them.